TOF workflow

noferini opened the pull request:
inserting TOF workflow:
Clustering moved from digitizer_workflow to tof workflow
track matching and calibration task in workflow
caching if TOF matrixes + refinmenets

noferini commented:
@chiarazampolli @shahor02 
Old pull request with Ruben's comment
https://github.com/AliceO2Group/AliceO2/pull/2207

shahor02 commented:
Hi @noferini ,
I still see some backup and link files, 
https://github.com/AliceO2Group/AliceO2/pull/2214/files#diff-14
https://github.com/AliceO2Group/AliceO2/pull/2214/files#diff-23

noferini commented:
> Hi @noferini ,
> I still see some backup and link files,
> https://github.com/AliceO2Group/AliceO2/pull/2214/files#diff-14
> https://github.com/AliceO2Group/AliceO2/pull/2214/files#diff-23

The backup files are a my mistake I will fix it.
About the links: it seems to me that git flagged with "->" the move from a directory to another directory, but they are not links. Am I wrong?

shahor02 commented:
@noferini probably yes, but the second link leads to https://github.com/AliceO2Group/AliceO2/blob/bc92664a44ba2e2d80dbef490d05c96ad2f9a56a/Detectors/GlobalTrackingWorkflow/tofworkflow/src/%23TOFClusterWriterSpec.cxx%23, now you have committed files ``#TOFClusterWriterSpec.cxx#`` and ``TOFClusterWriterSpec.cxx``

noferini commented:
Hi Ruben,

we implemented most of your comments.

- We removed one backup file

- We removed a couple of files (we don't need them):

DataFormats/Detectors/TOF/include/DataFormatsTOF/CompressedFormat.h

DataFormats/Detectors/TOF/include/DataFormatsTOF/Decoder.h

- We fixed data member names in our structs

- cable lengths moved to const (but not constexpr because we inizialize the values in a cxx file, because it is a very big matrix)

- default constructor used as suggested in Encoder class

- data member pointer initialized to null values

We didn't implement this suggestion:

- TOF DPL: suggestion "better to convert to task" all our Specs

We prefer to do that in the next pull request because functionality checks (more time) are needed to guarantee that results are correct.

One more question:
Do you still see problems with backup/link files?

noferini commented:
For your information, some examples on how to run our DPL:
_________________________________________________________
o2-tof-reco-workflow
input: TOF digits
output: TOF cluster, TOF matching info, TOF calib info
_________________________________________________________
o2-tof-reco-workflow --output-type clusters
input: TOF digits
output: TOF cluster
_________________________________________________________
o2-tof-reco-workflow --input-type clusters
input: TOF cluster
output: TOF matching info, TOF calib info
_________________________________________________________
o2-tof-reco-workflow --input-type clusters --output-type matching-info
input: TOF cluster
output: TOF matching info

shahor02 commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -665,29 +765,34 @@ void MatchTOF::doMatching(int sec)
+
+#ifdef _ALLOW_TOF_DEBUG_
           for (int ilabel = 0; ilabel < labelsTOF.size(); ilabel++) {
             LOG(DEBUG) << "TOF label " << ilabel << ": trackID = " << labelsTOF[ilabel].getTrackID() << ", eventID = " << labelsTOF[ilabel].getEventID() << ", sourceID = " << labelsTOF[ilabel].getSourceID();
```
note that the MCCompLabel has changed: the fake is not signaled by label<0 anymore, but by the special flag isFake, see #2171
You better use directly the label streaming method, then the complete info will be printed:
 ``LOG(DEBUG) << "TOF label " << ilabel << labelsTOF[ilabel] ;``

noferini replied:
LOG fixed. About the new fake flag I have to digest the change

shahor02 commented on Detectors/GlobalTracking/src/MatchTOF.cxx:
```diff
@@ -702,32 +807,37 @@ void MatchTOF::selectBestMatches()
+    // we want to store positive labels independently of how they are flagged from TPC,ITS people
+    //    o2::MCCompLabel labelTPC(abs(labelTPCor.getTrackID()), labelTPCor.getEventID(), labelTPCor.getSourceID());
+    //    o2::MCCompLabel labelITS(abs(labelITSor.getTrackID()), labelITSor.getEventID(), labelITSor.getSourceID());
     LOG(DEBUG) << "labelTPC: trackID = " << labelTPC.getTrackID() << ", eventID = " << labelTPC.getEventID() << ", sourceID = " << labelTPC.getSourceID();
```
idem

noferini replied:
done

noferini commented on DataFormats/Detectors/TOF/include/DataFormatsTOF/Encoder.h:
```diff
@@ -0,0 +1,61 @@
+  void setVerbose(bool val) { mVerbose = val; };
+
+  // benchmarks
+  double mIntegratedBytes = 0.;
```
now private

noferini commented on DataFormats/Detectors/TOF/src/Encoder.cxx:
```diff
@@ -0,0 +1,200 @@
+    auto BunchID = mUnion->CrateHeader.bunchID;
+    auto EventCounter = mUnion->CrateHeader.eventCounter;
+    auto DRMID = mUnion->CrateHeader.drmID;
+    std::cout << boost::format("%08x") % mUnion->Data
```
For the moment cout line commented and boost header no longer included

noferini commented on DataFormats/Detectors/TOF/src/Encoder.cxx:
```diff
@@ -0,0 +1,200 @@
+    unsigned char lastFilledFrame = 0;
+
+    /** loop over hits **/
+    /*
```
It has to be filled. I prefer to keep the comment for the moment

noferini commented on DataFormats/Reconstruction/include/ReconstructionDataFormats/CalibInfoTOF.h:
```diff
@@ -14,6 +14,8 @@
 #ifndef ALICEO2_CALIBINFOTOF_H
```
I moved all standalone tof classes in tof dir

noferini commented on Detectors/GlobalTracking/include/GlobalTracking/MatchTOF.h:
```diff
@@ -158,7 +167,13 @@ class MatchTOF
   void fillTOFmatchTree(const char* tname, int cacheTOF, int sectTOF, int plateTOF, int stripTOF, int padXTOF, int padZTOF, int cacheeTrk, int crossedStrip, int sectPropagation, int platePropagation, int stripPropagation, int padXPropagation, int padZPropagation, float resX, float resZ, float res, o2::dataformats::TrackTPCITS& trk, float intLength, float intTimePion, float timeTOF);
```
These come from the past. We will discuss how to clean this part of the code

noferini commented:
> Hi @noferini
> some comments from the previous review were not addressed, I've copied them here, see below (also added new ones).
> There still stray files:
> shahoian@alicers02:~/alice/O2$ find . -name "*.*" | grep '#'
> ./Detectors/GlobalTrackingWorkflow/tofworkflow/src/#TOFClusterWriterSpec.cxx#
Removed


> ./Detectors/GlobalTrackingWorkflow/tofworkflow/include/TOFWorkflow/#ClusterReaderSpec.h#
I didn't find this file (???)


> 
> Ok for the conversion to tasks, but what about moving the DPL devices to appropriate directories (#2207 (comment)) ?
> The GlobalTrackingWorkflow is for the specs and workflows which require (have dependency on) multiple detectors. All single detector devices, like TOFClusterizerSpec, ClusterReaderSpec, DigitReaderSpec etc. would belong to Detectors/TOF/workflow/ and referred by e.g. tof-reco-workflow.cxx in the Detectors/GlobalTrackingWorkflow (no need to put it in separate directory GlobalTrackingWorkflow/tofworkflow). Also for your tofworkflow/src/TrackReaderSpec.cxx the natural place would be where the ITSTPC matching is done and it should be renamed to e.g. TrackTPCITSReaderSpec.

Almost done. I need only to remove TrackReaderSpec by replacing it with the TrackTPCITSReaderSpec.

noferini commented on Detectors/GlobalTrackingWorkflow/tofworkflow/src/ClusterReaderSpec.cxx:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  mState = 2;
+  //  pc.services().get<ControlService>().readyToQuit(false);
```
done

noferini commented on Detectors/GlobalTrackingWorkflow/tofworkflow/src/DigitReaderSpec.cxx:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  mState = 2;
+  //  pc.services().get<ControlService>().readyToQuit(false);
```
done

noferini commented:
TrackReaderSpec replaced by TrackTPCITSReaderSpec.
Therefore, I think I covered all points.
Let us know if you think that something is still missing.
Cheers

chiarazampolli commented:
Hi Ruben, 
Yes, I am doing it incrementally, sorry, because we had also other comments that Francesco addressed yesterday. 
I will finish today.
Cheers,
Chiara

chiarazampolli commented:
Hello @noferini , @shahor02 ,

I committed some fixes, and open the PR on Francesco's branch.
One question for @shahor02 (or someone else, I don't know): I find the output of the o2checkcode a bit difficult to read (I hope I got all errors), isn't there a way to make it more explicit?

Chiara

noferini commented:
PR accepted

shahor02 commented:
Hi @chiarazampolli , the remaining checkode errors are not related to this PR, they come from the Framework. I think this PR can be merged, but I cannot do this if some of the tests are red, @ktf can.

ktf commented:
These is the usual clang bug. I need to remember how to mask it (and maybe we should bump clang for the code checker to a version which does not have the problem, now...).

ktf commented:
I assume you are happy with the 23 commits for this PR?

shahor02 commented:
Indeed, better to squash them, but doing this at merging will create misleading commit name, @noferini, could you do this manually?

noferini commented:
I'll squash commits this afternoon.
Cheers

noferini commented:
I'll squash commits this afternoon.
Cheers

noferini commented:
I did the squash (now 4 commits).
No diff with respect previous status.
cheers

shahor02 commented:
@noferini Thanks, due to the o2checkcode still failing for unrelated reasons I cannot merge it, @ktf, could you please do this?

ktf commented:
Yes, let me just wait a bit more for the O2/o2 test.
